pkg/ingester/limiter.go

```diff
@@ -106,22 +108,31 @@ func (l *Limiter) minNonZero(first, second int) int {
 }
 
 func (l *Limiter) convertGlobalToLocalLimit(globalLimit int) int {
-	if globalLimit == 0 {
+	if globalLimit == 0 || l.replicationFactor == 0 {
 		return 0
 	}
-	// todo: change to healthyInstancesInZoneCount() once
-	// Given we don't need a super accurate count (ie. when the ingesters
-	// topology changes) and we prefer to always be in favor of the tenant,
-	// we can use a per-ingester limit equal to:
-	// (global limit / number of ingesters) * replication factor
-	numIngesters := l.ring.HealthyInstancesCount()
 
-	// May happen because the number of ingesters is asynchronously updated.
-	// If happens, we just temporarily ignore the global limit.
+	zonesCount := l.ring.ZonesCount()
+	if zonesCount <= 1 {
+		return calculateLimitForSingleZone(globalLimit, l)
+	}
+
+	return calculateLimitForMultipleZones(globalLimit, zonesCount, l)
+}
+
+func calculateLimitForSingleZone(globalLimit int, l *Limiter) int {
+	numIngesters := l.ring.HealthyInstancesCount()
 	if numIngesters > 0 {
 		return int((float64(globalLimit) / float64(numIngesters)) * float64(l.replicationFactor))
 	}
+	return 0
+}
 
+func calculateLimitForMultipleZones(globalLimit, zonesCount int, l *Limiter) int {
+	ingestersInZone := l.ring.HealthyInstancesInZoneCount()
+	if ingestersInZone > 0 {
+		return int(float64(globalLimit) * float64(l.replicationFactor) * float64(zonesCount) / float64(ingestersInZone))
```

Comment:
```suggestion
		return int(float64(globalLimit) * float64(l.replicationFactor) / float64(zonesCount) / float64(ingestersInZone))
```
technically, I expect zones count to be equal to the replication factor, so we could simplify the expression to :
`int(float64(globalLimit) / float64(ingestersInZone))` because every zone receives exactly one replica of the data.